### PR TITLE
Make codegen tests compatible with extra inlining

### PR DESCRIPTION
**Pull request by tmiasko**

### CODE REVIEW

File: src/test/codegen/issue-37945.rs [hunk 1]
```diff
@@ -1,26 +1,34 @@
...
 
 use std::slice::Iter;
 
-// CHECK-LABEL: @is_empty_1
 #[no_mangle]
 pub fn is_empty_1(xs: Iter<f32>) -> bool {
-// CHECK-NOT: icmp eq float* {{.*}}, null
+// CHECK-LABEL: @is_empty_1(
+// CHECK-NEXT:  start:
+// CHECK-NEXT:    [[A:%.*]] = icmp ne i32* %xs.1, null
```
Comment (cuviper) [inline src/test/codegen/issue-37945.rs:18]: > Check that LLVM understands that `Iter` pointer is not null.

If we have `icmp ne ..., null`, it seems like we're failing the thing we meant to test, no?

I don't think we should try to compare exactly what LLVM IR comes out, but perhaps something like:

```
/// CHECK-NOT: icmp {{.*}}, null
```

**Review by cuviper — COMMENTED** [review]
It's not clear to me why we should be modifying tests without any corresponding compiler changes, but it does look like `issue-37945.rs` is problematic...

File: src/test/codegen/issue-37945.rs (hunk 1 above)
Comment (tmiasko) [reply src/test/codegen/issue-37945.rs:18]: > If we have `icmp ne ..., null`, it seems like we're failing the thing we meant to test, no?

It was fixed upstream by canonicalizing to assume.

**Comment by tmiasko** [comment]
> It's not clear to me why we should be modifying tests without any corresponding compiler changes

From my perspective those are bugs in tests. Fixes are mostly for my own convenience to be able to run tests with extra inlining. If you don't want to land some / any of those that seems fair.

File: src/test/codegen/issue-37945.rs (hunk 1 above)
Comment (cuviper) [reply src/test/codegen/issue-37945.rs:18]: By "it was fixed upstream", do you mean that will be in LLVM 12? Is there something we could cherry-pick?

My point is the test meant to check that LLVM knows it's not null, so we had a `CHECK-NOT: icmp eq ..., null` to that effect. Your change reveals that while we _were_ passing that test, we're getting `icmp ne ..., null` instead, which seems just as bad. So it seems like a real bug with or without your change.

Comment (tmiasko) [reply src/test/codegen/issue-37945.rs:18]: The fact that pointer is not-null is expressed through `llvm.assume` that follows, see next check.

Comment (cuviper) [reply src/test/codegen/issue-37945.rs:18]: If you mean these `assume`s, then the bug is that they're apparently not as effective as we want:

https://github.com/rust-lang/rust/blob/f5230fbf76bafd86ee4376a0e26e551df8d17fec/library/core/src/slice/iter/macros.rs#L142-L145

Comment (cuviper) [reply src/test/codegen/issue-37945.rs:18]: Hmm, ok, I see what you mean...

**Comment by cuviper** [comment]
OK, if CI is happy, I'm happy.

@bors r+

**Pull request merged**